library/timing.c

```diff
@@ -157,6 +158,26 @@ int mbedtls_timing_get_delay( void *data )
 
     return( 0 );
 }
+#else
```

Comment:
09e803ce0d7ec5acf29ddd9421f60e4a84fadb27 introduces a dummy implementation of timer functions. Now, when `defined(MBEDTLS_TIMING_C) && !defined(MBEDTLS_TIMING_ALT) && !defined(MBEDTLS_HAVE_TIME)` (did I get this condition right?), the functions `mbedtls_timing_get_timer`, `mbedtls_timing_set_delay` and `mbedtls_timing_get_delay` exist but they pretend that no time ever elapses.

What is the point of this dummy implementation? Why not make `MBEDTLS_TIMING_C` require `MBEDTLS_HAVE_TIME || MBEDTLS_TIMING_ALT`, since this seems to be necessary to implement the timing functions?

The dummy implementation breaks https://github.com/Mbed-TLS/mbedtls/pull/5582 and I'm trying to figure out how to fix this. Before #3624, in #5582, I made a configuration [`config-ccm-psk-dtls1_2.h`](https://github.com/Mbed-TLS/mbedtls/blob/6f160cab59897c23d49b8a6a704bcb834caf98ba/configs/config-ccm-psk-dtls1_2.h) with `MBEDTLS_TIMING_C` enabled but not `MBEDTLS_HAVE_TIME`, and I got `ssl-opt.sh` to pass. Since #3624, the test cases in `ssl-opt.sh` that use `udp_proxy … pack=50` [are failing](https://github.com/Mbed-TLS/mbedtls/pull/5582#issuecomment-1087752203), because the `pack` option of the proxy does not work with the dummy implementation of `mbedtls_timing_get_timer`. The `pack` options causes the proxy to accumulate packets until a certain amount of time has elapsed, but with the dummy implementation, this never happens, so the proxy never forwards any packet.

Currently, in `udp_proxy.c`, `opt.pack` is only accepted if `defined(MBEDTLS_TIMING_C)`. Should this change to `… && (defined(MBEDTLS_HAVE_TIME) || defined(MBEDTLS_TIMING_ALT))`?

Reply:
I think if we take the dummy implementation out there will be several compile failures - it's on my list to check, hopefully later today

Reply:
> I think if we take the dummy implementation out there will be several compile failures - it's on my list to check, hopefully later today

I have done that already - in the default config, `test_suite_ssl` and `test_suite_timing` will not compile, also `dtls_server`, `dtls_client` and `ssl_client2` from `programs` will fail to compile.

Reply:
I'll do a follow-up PR today and link it here.

Reply:
So we could just update `check_config.h` with something like

```
diff --git a/include/mbedtls/check_config.h b/include/mbedtls/check_config.h
index 1ced6e578..8aec5d3d7 100644
--- a/include/mbedtls/check_config.h
+++ b/include/mbedtls/check_config.h
@@ -59,6 +59,15 @@
 #error "MBEDTLS_DEPRECATED_WARNING only works with GCC and Clang"
 #endif
 
+#if defined(MBEDTLS_TIMING_C)
+#if !defined(MBEDTLS_HAVE_TIME) && !defined(MBEDTLS_TIMING_ALT)
+#error "MBEDTLS_TIMING_C without one of MBEDTLS_HAVE_TIME or MBEDTLS_TIMING_ALT"
+#endif
+#if defined(MBEDTLS_HAVE_TIME) && defined(MBEDTLS_TIMING_ALT)
+#error "Can't have both MBEDTLS_HAVE_TIME and MBEDTLS_TIMING_ALT"
+#endif
+#endif
+
 #if defined(MBEDTLS_HAVE_TIME_DATE) && !defined(MBEDTLS_HAVE_TIME)
 #error "MBEDTLS_HAVE_TIME_DATE without MBEDTLS_HAVE_TIME does not make sense"
 #endif
```

and then ensure we don't try to build with `MBEDTLS_TIMING_C` but without `MBEDTLS_HAVE_TIME_DATE` - that works now.

Is that likely to miss some build combinations we would like to cover?

We could also put in a build test with `MBEDTLS_TIMING_C`, without `MBEDTLS_HAVE_TIME_DATE`,  and with a dummy `MBEDTLS_TIMING_ALT` that just returns consecutive values each time it is called - would that help cover build combinations?

Reply:
If you take mbedtls-3.1 and build with default config minus `MBEDTLS_HAVE_TIME` and `MBEDTLS_HAVE_TIME_DATE` on a platform that actually has a working `time.h`, `make lib test` works but `make programs` doesn't.  The build of the library breaks in this same configuration if `time.h` is not available — which was the purpose of this PR.

Is it acceptable for us to break custom configurations that accidentally work because the developer didn't include `MBEDTLS_HAVE_TIME` in the configuration, but could have? If we do, at the very least, this should be prominently mentioned in the changelog entry.

Reply:
> If you take mbedtls-3.1 and build with default config minus MBEDTLS_HAVE_TIME and MBEDTLS_HAVE_TIME_DATE on a platform that actually has a working time.h, make lib test works but make programs doesn't

So you are removing any implementation (by saying "don't have time or date" and not providing `_ALT`) but leaving `MBEDTLS_TIMING_C` set, so this seems reasonable if we say that for `MBEDTLS_TIMING_C` you need either "have time/date" or `_ALT`

Reply:
> The build of the library breaks in this same configuration if time.h is not available

Err, I find that the library doesn't try to include a `time.h` file if I take `development` and just disable `MBEDTLS_HAVE_TIME` and `MBEDTLS_HAVE_TIME_DATE` - let me investigate further

Reply:
@tom-cosgrove-arm Before this PR (e.g. in 3.1), `platform.h` included `time.h` even when `MBEDTLS_HAVE_TIME` was disabled.

Reply:
@gilles-peskine-arm So I have double-checked: if I take a working Ubuntu build system, rename `time.h` out of the way, take Mbed TLS `development` and disable only `MBEDTLS_HAVE_TIME` and `MBEDTLS_HAVE_TIME_DATE`, the whole thing builds (as you say, that was the purpose of the PR). i.e. I can't reproduce your failure